src/Client/OfficeInterop/OfficeInterop.fs

```diff
@@ -1056,38 +1059,35 @@ let addBuildingBlock (excelTable:Table) (arcTable:ArcTable) (newBB:CompositeColu
 
             let mainColumNames =
                 // Get all cases of the union
-                Microsoft.FSharp.Reflection.FSharpType.GetUnionCases(typeof<CompositeHeader>)
-                |> Array.map (fun case -> case.Name)
-
-            let indices =
-                mainColumNames
-                |> Array.map (fun name ->
-                                headers
-                                |> Array.tryFindIndex (fun header -> header.StartsWith(name)))
-                                |> Array.where (fun index -> index.IsSome)
-                                |> Array.map (fun index -> index.Value)
-
-            if indices.Length > 0 then
+                CompositeHeader.Cases |> Array.map (fun (_, header) -> header)
+
+            let targetIndex =
+                [|
+                    for i in 0..headers.Length - 1 do
+                        let header = headers.[i]
+                        if mainColumNames |> Array.exists (fun cName -> header.StartsWith cName) then
+                            i
+                |]
+                |> Array.sortBy(fun index -> index)
+                |> Array.tryHead
 
-                let indices =
-                    indices
-                    |> Array.sortBy(fun index -> index)
+            if targetIndex.IsSome then
 
                 //The +1 makes sure, we add the column after the column we have currently chosen
-                excelIndex + (float indices.[0]) + 1.
+                excelIndex + (float) targetIndex.Value + 1.
 
             //The +1 makes sure, we add the column after the column we have currently chosen
-            else excelIndex + 1.
+            else AppendIndex
         else
             //Add the new building block to the end of the table
-            -1.
+            AppendIndex
 
     let headers =
         headerRange.values[0]
         |> List.ofSeq
         |> List.map (fun header -> header.ToString())
 
-    let bb = 
+    let _ = 
         buildingBlockCells
         |> List.mapi(fun i bbCell ->
```

Comment:
List.iteri, no binding. Otherwise we still instantiate a new list for performance loss

Reply:
fixed